src/core/loop.ts

```diff
@@ -0,0 +1,143 @@
+import type { Ref } from 'vue'
+import { ref } from 'vue'
+import { Clock, MathUtils } from 'three'
+import type { Fn } from '@vueuse/core'
+
+export type LoopStage = 'before' | 'render' | 'after'
+export interface LoopCallbackParams {
+  delta: number
+  elapsed: number
+  clock: Clock
+}
+export interface LoopCallback {
+  (params: LoopCallbackParams): void
+}
+
+export interface RendererLoop {
+  subscribersBefore: Map<number, Fn[]>
+  subscribersRender: Map<number, Fn[]>
+  subscribersAfter: Map<number, Fn[]>
+  loopId: string
+  register: (callback: LoopCallback, stage: LoopStage, index?: number) => void
+  start: () => void
+  stop: () => void
+  pause: () => void
+  resume: () => void
+  pauseRender: () => void
+  resumeRender: () => void
+  isActive: Ref<boolean>
+  isRenderPaused: Ref<boolean>
+}
+
+export function createRenderLoop(): RendererLoop {
+  const clock = new Clock(false)
+  const isActive = ref(false)
+  const isRenderPaused = ref(false)
+  let animationFrameId: number
+  const loopId = MathUtils.generateUUID()
+
+  const subscribersBefore = new Map()
+  const subscribersRender = new Map()
+  const subscribersAfter = new Map()
+
+  function registerCallback(callback: LoopCallback, stage: 'before' | 'render' | 'after', index = 0) {
+    let targetMap
+    if (stage === 'before') {
+      targetMap = subscribersBefore
+    }
+    else if (stage === 'render') {
+      targetMap = subscribersRender
+    }
+    else {
+      targetMap = subscribersAfter
+    }
+
+    if (stage === 'render' && index === 0) {
+      targetMap.set(index, [callback])
+    }
+    else {
+      if (!targetMap.has(index)) {
+        targetMap.set(index, [])
+      }
+      targetMap.get(index).push(callback)
+    }
+  }
+
+  function start() {
+    if (!isActive.value) {
+      clock.start()
+      isActive.value = true
+      loop()
+    }
+  }
+
+  function stop() {
+    if (isActive.value) {
+      clock.stop()
+      cancelAnimationFrame(animationFrameId)
+      isActive.value = false
+    }
+  }
+
+  function pause() {
+    clock.stop()
+    isActive.value = false
+  }
+
+  function resume() {
+    clock.start()
+    isActive.value = true
+  }
+
+  function pauseRender() {
+    isRenderPaused.value = true
+  }
+
+  function resumeRender() {
+    isRenderPaused.value = false
+  }
+
+  function loop() {
+    const delta = clock.getDelta()
+    const elapsed = clock.getElapsedTime()
+
+    const executeCallbacks = (subscribers: Map<number, Fn[]>) => {
+      Array.from(subscribers.keys())
+        .sort((a, b) => a - b) // Ensure numerical order
```

Comment:
For performance, it's probably best to avoid creating a new array and sorting on every tick of the loop, if possible.

Reply:
I agree. Maybe an array is the better choice over a Map since it doesn't have to be sorted every time.

Reply:
I also agree, I tried as well having the sorting indexes on an array every time you register a callback, but maybe not using Maps at all is a better option.

Also @andretchen0 provided me with another approach that he implemented on his PoC https://github.com/Tresjs/tres/pull/673#issuecomment-2096908045 with a `createPriorityEventHook`, I'm going to try that instead.

Thanks both for the  feedback 🙏🏻